### PR TITLE
Adding TGCN for nodeproppred

### PR DESCRIPTION
**Pull request by benjaminnNgo**
## Objective

Adding the implementation of [T-GCN](https://arxiv.org/abs/1811.05320) and a node property prediction example for T-GCN

## New scripts

- [tgm/nn/recurrent/tgcn.py](https://github.com/tgm-team/tgm/blob/79-tgcn-nodePropPrediction/tgm/nn/recurrent/tgcn.py)
- [examples/nodeproppred/tgcn.py](https://github.com/tgm-team/tgm/blob/79-tgcn-nodePropPrediction/examples/nodeproppred/tgcn.py)
- [test/integration/test_gclstm.py](https://github.com/tgm-team/tgm/blob/79-tgcn-nodePropPrediction/test/integration/test_gclstm.py)

### Command used:
```
python .\examples\nodeproppred\tgcn.py --dataset tgbn-trade --time-gran Y --batch-time-gran Y --epochs 100 --seed 720

python .\examples\nodeproppred\gclstm.py --dataset tgbn-trade --time-gran Y --batch-time-gran Y --epochs 100 --seed 720
```
- Cross Entropy Loss & validation NDCG over epochs:

test_tgcn

- Achieved test result (no early stopping): 0.2521 NDCG on `tgbn-trade` (`GCLSTM` achieved 0.2462 NDCG)

#### TODO for another PR:

- [ ] Add link prediction example for TGCN

### CODE REVIEW

**Comment by Jacob-Chmura** [comment]
> Not sure why the test check on GitHub failed

Safe to ignore these tests. There was an issue in the test cases due to auto-merging conflicts in some PRs that I pushed. I have adressed these and the fix should be in master shortly.

File: test/integration/test_tgcn.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+@pytest.mark.slurm(
+    resources=[
+        '--partition=main',
+        '--cpus-per-task=2',
+        '--mem=4G',
+        '--time=3:00:00',
+        '--gres=gpu:a100l:1',
+    ]
+)
+def test_gclstm_nodeprop_pred(slurm_job_runner, dataset):
```
Comment (Jacob-Chmura) [inline test/integration/test_tgcn.py]:
```suggestion
def test_tgcn_nodeprop_pred(slurm_job_runner, dataset):
```

File: examples/nodeproppred/tgcn.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
+r"""python -u gclstm.py --dataset tgbn-trade --time-gran r --batch-time-gran r
+python -u gclstm.py --dataset tgbn-genre --time-gran s --batch-time-gran D\
+example commands to run this script.
+"""
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/tgcn.py]:
```suggestion
"""
```
I think this can be removed

File: examples/nodeproppred/tgcn.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    def forward(
+        self,
+        batch: DGBatch,
+        node_feat: torch.Tensor,
+        h_0: torch.Tensor | None = None,
+    ) -> Tuple[torch.Tensor, ...]:
+        edge_index = torch.stack([batch.src, batch.dst], dim=0)
+        edge_weight = batch.edge_weight if hasattr(batch, 'edge_weight') else None  # type: ignore
+        z, h_0 = self.encoder(node_feat, edge_index, edge_weight, h_0)
+        z_node = z[batch.nid_to_idx[batch.node_ids]]  # type: ignore
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/tgcn.py]: After merging #60, the batch attribute to use is actually `global_to_local`.
```suggestion
        z_node = z[batch.global_to_local[batch.node_ids]]  # type: ignore
```

File: examples/nodeproppred/tgcn.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    def __init__(self, node_dim: int, embed_dim: int) -> None:
+        super().__init__()
+        self.recurrent = TGCN(in_channels=node_dim, out_channels=embed_dim)
+        self.linear = nn.Linear(embed_dim, embed_dim)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        edge_weight: torch.Tensor,
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/tgcn.py]:
```suggestion
        edge_weight: torch.Tensor | None, 
```

File: examples/nodeproppred/tgcn.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+from tgm.util.seed import seed_everything
+
+parser = argparse.ArgumentParser(
+    description='TGCN Example',
+    formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+)
+parser.add_argument('--seed', type=int, default=1337, help='random seed to use')
+parser.add_argument('--dataset', type=str, default='tgbn-genre', help='Dataset name')
+parser.add_argument('--device', type=str, default='cpu', help='torch device')
+parser.add_argument('--epochs', type=int, default=100, help='number of epochs')
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/tgcn.py]: Just double checking if we want to use `100` or `10` as default number of epochs

File: tgm/nn/recurrent/tgcn.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        H: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        f"""Forward pass.
+
+        Args:
+            X (PyTorch Tensor): Node features.
+            edge_index (PyTorch Long Tensor): Graph edge indices.
+            edge_weight (PyTorch Long Tensor, optional): Edge weight vector.
+            H (PyTorch Tensor, optional): Hidden state matrix for all nodes.
+            C (PyTorch Tensor, optional): Cell state matrix for all nodes.
```
Comment (Jacob-Chmura) [inline tgm/nn/recurrent/tgcn.py]:
```suggestion
```

Comment (Jacob-Chmura) [reply tgm/nn/recurrent/tgcn.py]: Unless we need this tensor in which case example should be updated

File: examples/nodeproppred/tgcn.py (hunk 5 above)
Comment (ntgbaoo) [reply examples/nodeproppred/tgcn.py]: I would say `100` epochs

**Comment by shenyangHuang** [comment]
test performance of 0.3654 is quite sensible. 
for `tgbn-trade` TGN achieves 0.374 test. So I would say the performance is quite reasonable. 
see [here](https://tgb.complexdatalab.com/docs/leader_nodeprop/#tgbn-trade) for the leaderboard.

File: tgm/nn/recurrent/tgcn.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    Reference: https://arxiv.org/abs/1811.05320
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        improved: bool = False,
+        cached: bool = False,
```
Comment (shenyangHuang) [inline tgm/nn/recurrent/tgcn.py:26]: when would you want to turn this one? what would be the benefit?

File: examples/nodeproppred/tgcn.py (hunk 5 above)
Comment (shenyangHuang) [reply examples/nodeproppred/tgcn.py]: for DTDG methods usually takes more epochs

**Review by shenyangHuang — COMMENTED** [review]
performance looks good.

File: tgm/nn/recurrent/tgcn.py (hunk 7 above)
Comment (ntgbaoo) [reply tgm/nn/recurrent/tgcn.py:26]: `improved` and `cached` are both used to configure GCNConv. By default, they are set to `False`. From what I am understanding from reading the [documentation from PyG](https://pytorch-geometric.readthedocs.io/en/2.5.2/generated/torch_geometric.nn.conv.GCNConv.html).

- If we set `improved = True` , the self-loops are added `A+2I` instead of `A+I`. This increases the weight of self-loops, giving each node’s own features more influence during aggregation.
- If we set `cached = True`, the layer computes the normalized adjacency matrix only once, and reuses it. This speeds up training and inference but is only suitable for transductive learning. Scenarios where `cached` can be set to `True` include tasks such as traffic forecasting, where graph structure is assumed to be static and only node features are dynamic. `cached` should not be set to `False` when graph structure dynamically changes.

Thanks for pointing this out. These are all useful information that we should include in the comment section.

**Comment by benjaminnNgo** [comment]
I added further information about `cached` and `improved` in the comment section and increased the default number of epochs as you suggested, @shenyangHuang. Please take a look whenever you have a chance.

File: tgm/nn/recurrent/tgcn.py (hunk 7 above)
Comment (shenyangHuang) [reply tgm/nn/recurrent/tgcn.py:26]: yeah thanks for adding the comments here, much more clear now

**Review by Jacob-Chmura — APPROVED** [review]

**Review by shenyangHuang — APPROVED** [review]

**Comment by shenyangHuang** [comment]
The new performance makes sense to me, I got similar range from GCN yesterday for nodeproppred.
